components/script/dom/htmlimageelement.rs

```diff
@@ -882,6 +931,182 @@ impl HTMLImageElement {
+                // Step 15
+                self.finish_reacting_to_environment_change(selected_source, generation, selected_pixel_density);
+            }
+
+            Ok(ImageOrMetadataAvailable::MetadataAvailable(m)) => {
+                self.process_image_response_for_environment_change(ImageResponse::MetadataLoaded(m),
+                                                                   selected_source, generation, selected_pixel_density);
+            }
+
+            Err(ImageState::Pending(id)) => {
+                add_cache_listener_for_element(image_cache.clone(), id, self,
+                                               selected_source.to_string(), selected_pixel_density);
+            }
+
+            Err(ImageState::LoadError) => {
+                self.process_image_response_for_environment_change(ImageResponse::None,
+                                                                   selected_source, generation, selected_pixel_density);
+            }
+
+            Err(ImageState::NotRequested(id)) => {
+                add_cache_listener_for_element(image_cache, id, self, selected_source.to_string(),
+                                               selected_pixel_density);
+                self.fetch_request(&img_url, id);
+            }
+        }
+    }
+
+    /// Step 15 for <https://html.spec.whatwg.org/multipage/#img-environment-changes>
+    fn finish_reacting_to_environment_change(&self, src: DOMString, generation: u32, selected_pixel_density: f64) {
+        let this = Trusted::new(self);
+        let window = window_from_node(self);
+        let src = src.to_string();
+        let _ = window.dom_manipulation_task_source().queue(
+            task!(image_load_event: move || {
+                let this = this.root();
+                let relevant_mutation = this.generation.get() != generation;
+                let mut pending_request = this.pending_request.borrow_mut();
+
+                // Step 15.1
+                if relevant_mutation {
+                    this.abort_request(State::Unavailable, ImageRequestPhase::Pending);
+                    return;
+                }
+
+                // Step 15.2
+                *this.last_selected_source.borrow_mut() = Some(DOMString::from_string(src));
+                pending_request.current_pixel_density = Some(selected_pixel_density);
+
+                // Step 15.3
+                pending_request.state = State::CompletelyAvailable;
+
+                // Step 15.4
+                // Already a part of the list of available images due to Step 14
+
+                // Step 15.5
+                mem::swap(&mut this.current_request.borrow_mut(), &mut pending_request);
+                this.abort_request(State::Unavailable, ImageRequestPhase::Pending);
+
+                // Step 15.6
+                this.upcast::<Node>().dirty(NodeDamage::OtherNodeDamage);
+
+                // Step 15.7
+                this.upcast::<EventTarget>().fire_event(atom!("load"));
```

Comment:
This is one place where we fire an event with an outstanding borrow (pending_request).

Reply:
Two possible solutions:
1) remove the local variable and use self.pending_request.borrow()/self.pending_request.borrow_mut() everywhere
2) add a block around all the code that uses the borrowed value ({}) so that it is dropped before we call fire_event